README
NAME

    MooseX::Role::UserDaemon - Simplify writing of user space daemons

VERSION

    version 0.05

SYNOPSIS

    In your module:

      package YourApp;
      use Moose;
      with qw(MooseX::Role::UserDaemon);
    
      # MooseX::Role::UserDaemon requires the consuming class to implement main()
      sub main {
        my ($self) = @_;
    
        # It is the responsibility of the consuming class to capture INT signals
        # to allow for graceful shutdown of the app.
        # In addition the HUP signal should be caught and used for config reload.
        my $run = 1;
        local $SIG{'TERM'} = local $SIG{'INT'} = sub { $run = 0; };
    
        FOREVER_LOOP:
        while ($run) {
          ...
        }
    
        # It is recomended that main() return '0 but true' on success.
        # the return value of main is feed directly to exit()
        return '0 but true';
      }

    In your script:

      use YourApp;
      my $app = YourApp->new;
    
      exit $app->run unless caller 0;

    On the commanline:

      Start your app
      $ yourapp.pl start
        Starting...
    
      Check your status
      $ yourapp.pl status
        Running with PID: ...
    
      Stop your app
      $ yourapp.pl stop
        Stopping PID: ...

    Or preferably in combination with MooseX::SimpleConfig and/or
    MooseX::Getopt In your module:

      package YourApp;
      use Moose;
    
      # Enable use of configfile and commandline parameters as well
      with qw(MooseX::Role::UserDaemon MooseX::SimpleConfig MooseX::Getopt);
    
      # '+configfile' Only required when using MooseX::SimpleConfig
      has '+configfile' => (
        is      => 'ro',
        isa     => 'Str',
        default => sub {
          File::Spec->catdir( File::HomeDir->my_home, '.yourapp',
            'yourapp.conf' );
        },
        documentation => 'Use custom configfile.',   # Getopt use this description
      );
    
      # MooseX::Role::UserDaemon requires the consuming class to implement main()
      sub main {
        my ($self) = @_;
    
        # the user have to implement capturing signals and exiting.
        my $run = 1;
        local $SIG{'TERM'} = local $SIG{'INT'} = sub { $run = 0; };
    
        FOREVER_LOOP:
        while ($run) {
          sleep 1;
        }
          
        # It is recomended that main() return '0 but true' on success.
        # the return value of main is feed directly to exit()
        return '0 but true';
      }

    In your script:

      use YourApp;
    
      # use new_with_options when using Getopt/SimpleConfig
      my $app = YourApp->new_with_options;
    
      exit $app->run unless caller 0;

    On the commanline:

      Start your app
      $ yourapp.pl start
        Starting...
    
      Check your status
      $ yourapp.pl status
        Running with PID: ...
    
      Stop your app
      $ yourapp.pl stop
        Stopping PID: ...

DESCRIPTION

    This module aims to simplify implementation of daemons and apps ment to
    be run for normal users. Not system wide services or servers.

    This module is not suited for implementing daemons running as root or
    other system users.

    When using this role your script will by default:

    1. Create a hidden folder in the users home directory with the same
    name as the script itself. So YourApp.pl will create a directory
    ~/.yourapp.pl

    2. chdir to this directory.

    3. Daemonize by double forking.

    4. Create a lockfile named lock and place a flock on the file. This
    lock will be in place until the app shuts down.

    5. Create a pidfile named pid.

    6. run the main() subroutine.

    Five commands are implemented by default: status, start, stop, restart,
    reload. More information on these can be found in the methods section.

ATTRIBUTES

 _name

    String. Defaults to script name, is used for setting a application
    folder name.

 _valid_commands

    Regexp. Default is qr/\A(status|start|stop|reload|restart)\z/xms.
    Whitelist methods which can be called from the command line.

 timeout

    Integer. Default is 5. How much time in seconds it's expected to take
    after shutting down the app by sending a TERM singal. This is used by
    stop to avoid waiting forever for the app to shut down.

 foreground

    Integer. Default is 0. If set to 1 the app will not daemonize/fork or
    redirect STD* to /dev/null.

 basedir

    String. Default to an application folder in the users home directory.
    The app will chdir to this location during startup. This is where we
    will place lockfile, pidfile and other application files.

 lockfile

    String. Default to /basedir/lockfile

 _lockfile_fh

    Filehandle. Used for holding the lockfile filehandle while the app is
    running.

 pidfile

    String. Default to /basedir/pid

METHODS

 run

    run() will determine which command was issued to the script, defaults
    to start if no command is given. By default the valid commands are:
    status, start, stop, restart, reload

    New commands can be added by the consuming class, it which case the
    attribute _valid_commands needs to be updated for run() to allow the
    command to be executed. _valid_commands is a RegexpRef and the default
    value is: qr/\A(status|start|stop|reload|restart)\z/

    You can set your own _valid_commands in the consuming class, to allow
    for custom commands like this:

      has '+_valid_commands' => (
        default => sub {
          return qr/\A(status|start|stop|restart|custom_command)\z/xms
        },
      );

 status

    Checks if the app is running, print status to STDOUT.

 start

    start() call main() after checking that it is not running and after
    forking (unless foreground mode is enabled).

 stop

    stop() issues a TERM signal to the PID listed in the pidfile. It is up
    to the author to trap this signal and end the application in an orderly
    fashion.

 restart

    restart() simply call stop(), wait for the app to stop and call
    start().

 reload

    reload() issues a HUP signal to the PID listed in the pidfile. It is up
    to the author to trap this signal and do the appropriate thing, usualy
    to reload configuration files.

AUTHOR

    Tore Andersson <tore.andersson@gmail.com>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2014 by Tore Andersson.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
==================================================
Changes from 2013-05-10 00:00:00 +0000 to present.
==================================================

-----------------------------------------
version 0.05 at 2014-05-10 17:00:40 +0000
-----------------------------------------

  Change: 31c4b352a6a5f9e12651ebe7274a6fa6de511bd9
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-05-10 18:59:17 +0000

    Extract basename first to have the name be consistent regardless of
    how it was called. 

  Change: a829b8482a3af30718f32e067443c14579ecd7e4
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-05-10 00:16:33 +0000

    Removed redundant POD 

  Change: 2f441ff66b4edecf2febe3604925cee1d7266deb
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-05-09 00:28:13 +0000

    updated README.pod 

  Change: 2697b3934ff2e82ce55c331afae63074bb0e2ef4
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-05-09 00:26:00 +0000

    Reworked unit tests, improved coverage 

  Change: 1db5f0ae784d4430e64895bb7879a8210ec6adc6
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-04-27 01:59:22 +0000

    Added more unit tests, also changed code to make testing easier and
    fixing bugs 

  Change: 2e5398c96a09c0b085a0a262b8e62c8d52288953
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-04-25 23:59:30 +0000

    fixed unit tests after making lockfile and pidfile read-only 

  Change: a9a40db3edc49f431651df967cf312abf05da4f1
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-04-25 00:37:24 +0000

    replaced sleep loop with alarm in stop() (giving up trying to support
    windows). Also no longer protecting the start stop and reload command
    against excessive number of requests. Lastly a few changes to make
    perlcritic happier. 

  Change: d0447bcdb29bb60368e7d2114e012c5dcf889bc3
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-04-23 22:34:58 +0000

    added more unit tests 

  Change: c8277f13a380d30748b463555589d212b3384d9d
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-04-23 22:34:14 +0000

    Aviod exiting prematurly when usage text is envoked by invalid
    commands + stylistic changes 

  Change: 1e9a49697243292893b3b4082b4b3e8210355f71
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-04-16 21:06:02 +0000

    Added fixes to be able to run under tainmode 

  Change: 03ab92ba1419eec3e2b3963847be823c712d3523
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-03-08 13:44:24 +0000

    No longer remove lockfile after unlocking 

  Change: 7782956db907d69aae36e5e2a3dd040c2998e5bf
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-03-08 11:02:56 +0000

    Now reopen STD* to /dev/null after the second fork, and change a few
    comments 

  Change: fde02bbc77e336b18b824576dbc5b6a5c8b939fe
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-03-02 20:00:49 +0000

    stylistic changes 

  Change: 8fee7d93805b8cefba00634cd22e8996a8f3518a
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-03-02 19:35:59 +0000

    added .gitignore to the repo 

  Change: 908cee5900ebda023a361947b7e717a001dc4919
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-02-13 00:11:32 +0000

    Added protection for the sleep call in stop(), added _init_fh() to
    remove duplicate code, and some other minor changes 

  Change: c52e3cd91bdf4689e98d9789d4fca21ebee9384a
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-02-03 23:44:09 +0000

    fixed long qw() string 

  Change: ec32345e3e5f3fd197c8237f1df07342a7b7d7fa
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-02-03 23:26:10 +0000

    Added unit test for run() 

  Change: 7c08bb879087f7265377825bfeccb344210d3318
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-02-02 17:22:48 +0000

    Fixed make_path to File::Path::make_path 

  Change: 000d1fa1fdead2ae8100c53ae86ba396c987ac88
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-02-02 16:52:05 +0000

    added/changed a few comments 

  Change: 223b7c1360b280011b87be283941726b81728ab6
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-02-02 15:52:35 +0000

    Minimize use of imports, updated dist.ini 

  Change: 02dda5ac321481dc066e47925f625cc76fc4afed
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-02-01 20:54:30 +0000

    Added/changed more documentation 

  Change: 2e6f7958b0271b936378aaa586446d4f73c5a3e7
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-01-28 22:53:59 +0000

    Added more documentation and updated POD 

  Change: 962bd285c0e518c00eadb81f4f4647da9e95cfe9
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-01-28 22:52:47 +0000

    Change from hardcoded filepaths to File::Spec 

  Change: 5250213c49b53082e6cef3be7f76971c4e6c3bb6
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-01-20 23:09:14 +0000

    Yet another minor POD fix 

  Change: 5472a3fe2d0245a9761bea8625adce136fc0abbc
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-01-20 23:04:38 +0000

    Update POD and readme 

  Change: e63493673c0334afd981e0822054ebb5fa23c330
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-01-19 23:58:54 +0000

    Updated README.pod with dzil 

  Change: 0d6e9bd25b475e28caeb043b3b9291bf73cccfa6
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-01-19 23:55:54 +0000

    Minor changes to POD 

  Change: 084a2559a62a4d46332f4b96fb33362babf042a7
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-01-19 23:29:48 +0000

    Converting to dist::zilla for release management 

  Change: beae36f232491905adf667b7ea5f10d6eebbc7dd
  Author: Tore Andersson <tore.andersson@gmail.com>
  Date : 2014-01-01 19:47:06 +0000

    Updated README, MANIFEST and a few comments in the code 


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.016.
use strict;
use warnings;

use 5.014;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Simplify writing of user space daemons",
  "AUTHOR" => "Tore Andersson <tore.andersson\@gmail.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "MooseX-Role-UserDaemon",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "MooseX::Role::UserDaemon",
  "PREREQ_PM" => {
    "English" => 0,
    "Fcntl" => 0,
    "File::Basename" => 0,
    "File::HomeDir" => 0,
    "File::Path" => 0,
    "File::Spec" => 0,
    "Moose::Role" => 0,
    "POSIX" => 0,
    "autodie" => 0,
    "namespace::autoclean" => 0
  },
  "TEST_REQUIRES" => {
    "Data::Dumper" => 0,
    "File::Temp" => 0,
    "Moose" => 0,
    "MooseX::Getopt" => 0,
    "Readonly" => 0,
    "Test::More" => 0,
    "Test::Most" => 0,
    "Test::Output" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.05",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Data::Dumper" => 0,
  "English" => 0,
  "Fcntl" => 0,
  "File::Basename" => 0,
  "File::HomeDir" => 0,
  "File::Path" => 0,
  "File::Spec" => 0,
  "File::Temp" => 0,
  "Moose" => 0,
  "Moose::Role" => 0,
  "MooseX::Getopt" => 0,
  "POSIX" => 0,
  "Readonly" => 0,
  "Test::More" => 0,
  "Test::Most" => 0,
  "Test::Output" => 0,
  "autodie" => 0,
  "namespace::autoclean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/MooseX/Role/UserDaemon.pm
package MooseX::Role::UserDaemon;

# ABSTRACT: Simplify writing of user space daemons

use 5.014;
use Moose::Role;
use autodie;
use English qw(-no_match_vars);
use Fcntl qw(:flock);
use File::Basename ();
use File::HomeDir  ();
use File::Spec     ();
use File::Path     ();
use POSIX          ();
use namespace::autoclean;

BEGIN {
  our $VERSION = '0.05'; # VERSION
}

{
  requires 'main';

  has '_name' => (
    is      => 'ro',
    isa     => 'Str',
    lazy    => 1,
    default => sub {    # The program name should not be to exotic.

      # Aviod modifying $PROGRAM_NAME directly
      my $program_name = File::Basename::fileparse $PROGRAM_NAME;

      # First replace exotic chars with _;
      $program_name =~ s/[^\w.,_ -]+/_/g;

      # Then capture to pass taintmode.
      ($program_name) = $program_name =~ /\A([\w.,_ -]+)\z/
        or die 'program name contain invalid characters.';

      return $program_name;
    },
  );

  has '_valid_commands' => (
    is       => 'ro',
    isa      => 'RegexpRef',
    lazy     => 1,
    default  => sub {qr/\A(status|start|stop|reload|restart)\z/},
    init_arg => undef,
  );

  has 'timeout' => (
    is      => 'ro',
    isa     => 'Int',
    lazy    => 1,
    default => 5,
    documentation =>
      '--timeout=n, default = 5, time in seconds to wait for the daemon to exit.',
  );

  has 'foreground' => (
    is      => 'ro',
    isa     => 'Int',
    lazy    => 1,
    default => 0,
    documentation =>
      '--foreground=1 will run the app in the foreground instead of daemonizing it.',
  );

  has 'basedir' => (
    is            => 'ro',
    isa           => 'Str',
    lazy_build    => 1,
    documentation => '--basedir="/custom/path", Use custom base directory.',
  );

  has 'lockfile' => (
    is            => 'ro',
    isa           => 'Str',
    lazy_build    => 1,
    documentation => '--basedir=/path/to/file, Use custom lockfile.',
  );

  has 'pidfile' => (
    is            => 'ro',
    isa           => 'Str',
    lazy_build    => 1,
    documentation => '--pidfile=/path/to/file, Use custom pidfile.',
  );

  has '_lock_fh' => (
    is       => 'rw',
    isa      => 'Maybe[FileHandle]',
    init_arg => undef,
    clearer  => 'clear_lock_fh',
  );

  sub _build_basedir {
    my ($self) = @_;
    return File::Spec->catdir( File::HomeDir->my_home,
      lc( q{.} . $self->_name ) );
  }

  sub _build_lockfile {
    my ($self) = @_;
    return File::Spec->catdir( $self->basedir, 'lock' );
  }

  sub _build_pidfile {
    my ($self) = @_;
    return File::Spec->catdir( $self->basedir, 'pid' );
  }

  # Write PID file if supplied
  around 'main' => sub {
    my ( $orig, $self ) = @_;

    # Abort if no pidfile has been specified
    return if !$self->pidfile;

    # Failed to write pid
    return 0 if !$self->_write_pid;

    my $rc = $self->$orig;

    # Failed to remove pidfile after execution
    return 0 if !$self->_delete_pid;

    return $rc;
  };

  # Write lock file if supplied
  around 'main' => sub {
    my ( $orig, $self ) = @_;

    # Abort if no lockfile has been specified
    return if !$self->lockfile;

    # Failed to establish a lock
    return 0 if !$self->_lock;

    # run main
    my $rc = $self->$orig;

    # Failed to remove lock
    return 0 if !$self->_unlock;

    return $rc;
  };

  sub _init_fh {
    my ( $self, $mode, $filename ) = @_;

    # Create leading directories if missing.
    if ( !-e $filename ) {
      File::Path::make_path($filename);
      rmdir $filename;
    }

    # Open filehandle
    open my ($filehandle), $mode, $filename;
    return $filehandle;
  }

  sub _lockfile_is_valid {
    my ($self) = @_;

    if ( !-f $self->lockfile ) {
      warn 'lockfile is not a regular file.';
      return;
    }

    if ( !-z $self->lockfile ) {
      warn 'lockfile is not a empty file.';
      return;
    }

    if ( !-w $self->lockfile ) {
      warn 'lockfile is not writeable by the current process.';
      return;
    }

    return 1;
  }

  sub _lock {
    my ($self) = @_;

    die 'Must specify a path to be used as a lockfile.'
      if !$self->lockfile;

    die 'A lockfile already exists but it is not an empty writable file.'
      if -e $self->lockfile && !$self->_lockfile_is_valid;

    # Finally open the file and place a lock on it
    my $lock_fh = $self->_init_fh( '>>', $self->lockfile );
    flock $lock_fh, LOCK_EX | LOCK_NB or return;

    # Maintain the lock troughout the runtime of the app. Store the FH.
    $self->_lock_fh($lock_fh);

    return 1;
  }

  sub _unlock {
    my ($self) = @_;

    die 'Trying to unlock a non-existing lockfile filehandle.'
      if !$self->_lock_fh;

    my $close_rc = close $self->_lock_fh;

    if ( !$close_rc ) {
      warn "Failed to close lockfile filehandle: $ERRNO";
    }
    else {
      $self->clear_lock_fh;
    }

    return $close_rc;
  }

  sub _pidfile_is_valid {
    my ($self) = @_;

    if ( !-f $self->pidfile ) {
      warn 'pidfile is not a file';
      return;
    }

    if ( !-r $self->pidfile ) {
      warn 'pidfile is not readable';
      return;
    }

    if ( !-w $self->pidfile ) {
      warn 'pidfile is not writeable';
      return;
    }

    return 1;
  }

  sub _write_pid {
    my ($self) = @_;

    die 'Must specify a path to be used as a pidfile.'
      if !$self->pidfile;

    die 'A pidfile already exist, but is not a regular writable file.'
      if -e $self->pidfile && !$self->_pidfile_is_valid;

    # write the actual file
    {
      ## no critic (ProhibitLocalVars)
      local $OUTPUT_AUTOFLUSH = 1;

      ## use critic
      my $pid_fh = $self->_init_fh( '>', $self->pidfile );
      print {$pid_fh} $PID;
      close $pid_fh;
    }

    return 1;
  }

  sub _read_pid {
    my ($self) = @_;

    die 'Must specify a path to be used as a pidfile.'
      if !$self->pidfile;

    die 'A pidfile already exist, but is not a regular file.'
      if -e $self->pidfile && !$self->_pidfile_is_valid;

    open my $pid_fh, '<', $self->pidfile;
    my $daemon_pid = do { local $INPUT_RECORD_SEPARATOR = undef; <$pid_fh> };
    ($daemon_pid) = $daemon_pid =~ /\A(\d+)\z/
      or die 'pidfile contained something other than digits.';
    close $pid_fh;

    return $daemon_pid;
  }

  sub _delete_pid {
    my ($self) = @_;

    die 'Must specify a path to be used as a pidfile.'
      if !$self->pidfile;

    die 'pidfile does not exist'
      if !-e $self->pidfile;

    die 'pidfile is not a regular file'
      if !-f $self->pidfile;

    die 'pidfile is not writable'
      if !-w $self->pidfile;

    return unlink $self->pidfile;
  }

  # Check if exists a locked version of the lockfile already
  sub _is_running {
    my ($self) = @_;

    # return false if lockfile is not in use
    return if !$self->lockfile || !-e $self->lockfile;

    if ( $self->_lock ) {    # Not running
      $self->_unlock;
      return;
    }

    return 1;                # running
  }

  sub _daemonize {
    my ($self) = @_;

    # Fork once
    defined( my $pid1 = fork ) or die "Can’t fork: $ERRNO";
    return '0 but true' if $pid1;    # Original parent exit

    # Become session leader
    POSIX::setsid
      or die "Unable to to become session leader: $ERRNO";

    # Fork twice
    defined( my $pid2 = fork ) or die "Can’t fork: $ERRNO";
    exit if $pid2;                   # Intermediate parent exit

    # Redirect STD* to /dev/null
    open STDIN,  '<',  File::Spec->devnull;
    open STDOUT, '>>', File::Spec->devnull;
    open STDERR, '>>', File::Spec->devnull;

    # Child returns false.
    return;
  }

  sub status {
    my ($self) = @_;

    say $self->_is_running
      ? 'Running with PID: ' . $self->_read_pid
      : 'Not running.';

    return '0 but true';
  }

  sub start {
    my ($self) = @_;

    return $self->status if $self->_is_running;

    say 'Starting...';

    # Return with output of main if foreground mode is enabled
    return $self->main
      if $self->foreground;

    # Else Daemonize
    my $daemonize_rc = $self->_daemonize;

    # Original parent returns
    return $daemonize_rc if defined $daemonize_rc;

    # Child will run main and exit when running as daemon
    exit $self->main;
  }

  sub stop {
    my ($self) = @_;

    if ( !$self->_is_running ) {
      say 'Process not running, nothing to stop.';
      return '0 but true';
    }

    if ( !$self->pidfile || !-e $self->pidfile ) {
      say 'No pidfile, not able to identify process.';
      return 0;    # Return 0 to please both unit tests and exit
    }

    # Get process id
    my $pid = $self->_read_pid;

    eval { kill 'TERM', $pid };
    if ($EVAL_ERROR) {
      warn 'Not able to issue kill signal.';
      return 0;
    }
    else {
      say "Stopping PID: $pid...";
    }

    eval {
      local $SIG{'ALRM'} = sub { die "Timed out waiting for exit\n"; };

      my $lockfile_fh = $self->_init_fh( '+<', $self->lockfile );
      if ( !flock( $lockfile_fh, LOCK_EX | LOCK_NB ) ) {

        #say "Lockfile still blocked, waiting\n";
        alarm $self->timeout;
        flock( $lockfile_fh, LOCK_EX );
        alarm 0;
      }
      close $lockfile_fh;
    };
    return $EVAL_ERROR
      ? 0
      : '0 but true';
  }

  sub restart {
    my ($self) = @_;

    # Start a new if stop is successful
    return $self->start if $self->stop;

    # Stop failed:
    say 'Failed to stop process, restart aborted.';

    # Return 0 to please both unit tests and exit
    return 0;
  }

  sub reload {
    my ($self) = @_;

    # Unable to signal process if no pidfile
    return 0 if !$self->pidfile || !$self->_is_running;

    # Get the process id
    my $pid = $self->_read_pid;

    # Signal the process
    my $kill_rc = eval { kill 'HUP', $pid };
    if ($EVAL_ERROR || !$kill_rc) {
      say "Failed to signal PID: $pid.";
      return 0;    # Return 0 to please both unit tests and exit
    }

    say "PID: $pid, was signaled to reload.";
    return '0 but true';
  }

  sub run {
    my ($self) = @_;

    # Get the command to run.
    my $command
      = $self->can('extra_argv') && ref( $self->extra_argv ) eq 'ARRAY'
      ? shift $self->extra_argv    # If MooseX::Getopt is in use.
      : shift @ARGV;               # Else get it from @ARGV

    # Default to start.
    $command ||= 'start';

    # Validate that mode is valid/approved
    ($command) = $command =~ $self->_valid_commands or do {
      say $self->can('usage')
        && ref( $self->usage )
        && $self->usage->can('text')
        ? $self->usage->text              # If MooseX::Getopt is in use
        : 'The command is not valid.';    # else default to a simple message
      return 0;
    };

    # Create base dir if none exists.
    if ( !-e $self->basedir && !File::Path::make_path( $self->basedir ) ) {
      say 'Failed to create basedir.';
      return 0;
    }

    # Change to running dir, fail if base dir is not a directory
    if ( !-d $self->basedir || !chdir $self->basedir ) {
      say 'Failed to enter basedir.';
      return 0;
    }

    # Run!
    return $self->$command;
  }
}

no Moose::Role;
1;

__END__

=pod

=encoding UTF-8

=head1 NAME

MooseX::Role::UserDaemon - Simplify writing of user space daemons

=head1 VERSION

version 0.05

=head1 SYNOPSIS

In your module:

  package YourApp;
  use Moose;
  with qw(MooseX::Role::UserDaemon);

  # MooseX::Role::UserDaemon requires the consuming class to implement main()
  sub main {
    my ($self) = @_;

    # It is the responsibility of the consuming class to capture INT signals
    # to allow for graceful shutdown of the app.
    # In addition the HUP signal should be caught and used for config reload.
    my $run = 1;
    local $SIG{'TERM'} = local $SIG{'INT'} = sub { $run = 0; };

    FOREVER_LOOP:
    while ($run) {
      ...
    }

    # It is recomended that main() return '0 but true' on success.
    # the return value of main is feed directly to exit()
    return '0 but true';
  }

In your script:

  use YourApp;
  my $app = YourApp->new;

  exit $app->run unless caller 0;

On the commanline:

  Start your app
  $ yourapp.pl start
    Starting...

  Check your status
  $ yourapp.pl status
    Running with PID: ...

  Stop your app
  $ yourapp.pl stop
    Stopping PID: ...

Or preferably in combination with MooseX::SimpleConfig and/or MooseX::Getopt
In your module:

  package YourApp;
  use Moose;

  # Enable use of configfile and commandline parameters as well
  with qw(MooseX::Role::UserDaemon MooseX::SimpleConfig MooseX::Getopt);

  # '+configfile' Only required when using MooseX::SimpleConfig
  has '+configfile' => (
    is      => 'ro',
    isa     => 'Str',
    default => sub {
      File::Spec->catdir( File::HomeDir->my_home, '.yourapp',
        'yourapp.conf' );
    },
    documentation => 'Use custom configfile.',   # Getopt use this description
  );

  # MooseX::Role::UserDaemon requires the consuming class to implement main()
  sub main {
    my ($self) = @_;

    # the user have to implement capturing signals and exiting.
    my $run = 1;
    local $SIG{'TERM'} = local $SIG{'INT'} = sub { $run = 0; };

    FOREVER_LOOP:
    while ($run) {
      sleep 1;
    }
      
    # It is recomended that main() return '0 but true' on success.
    # the return value of main is feed directly to exit()
    return '0 but true';
  }

In your script:

  use YourApp;

  # use new_with_options when using Getopt/SimpleConfig
  my $app = YourApp->new_with_options;

  exit $app->run unless caller 0;

On the commanline:

  Start your app
  $ yourapp.pl start
    Starting...

  Check your status
  $ yourapp.pl status
    Running with PID: ...

  Stop your app
  $ yourapp.pl stop
    Stopping PID: ...

=head1 DESCRIPTION

This module aims to simplify implementation of daemons and apps ment to be run
for normal users. Not system wide services or servers.

B<< This module is not suited for implementing daemons running as root
or other system users. >>

When using this role your script will by default:

1. Create a hidden folder in the users home directory with the same
name as the script itself. So YourApp.pl will create a directory ~/.yourapp.pl

2. C<< chdir >> to this directory.

3. Daemonize by double forking.

4. Create a lockfile named lock and place a C<< flock >> on the file.
This lock  will be in place until the app shuts down.

5. Create a pidfile named pid.

6. run the C<< main() >> subroutine.

Five commands are implemented by default: status, start, stop, restart,
reload. More information on these can be found in the methods section.

=head1 ATTRIBUTES

=head2 _name

String. Defaults to script name, is used for setting a application folder name.

=head2 _valid_commands

Regexp. Default is C<< qr/\A(status|start|stop|reload|restart)\z/xms >>.
Whitelist methods which can be called from the command line.

=head2 timeout

Integer. Default is 5. How much time in seconds it's expected to take after
shutting down the app by sending a C<< TERM >> singal. This is used by
C<< stop >> to avoid waiting forever for the app to shut down.

=head2 foreground

Integer. Default is 0. If set to 1 the app will not daemonize/fork or redirect
STD* to /dev/null.

=head2 basedir

String. Default to an application folder in the users home directory. The app
will C<< chdir >> to this location during startup. This is where we will place 
lockfile, pidfile and other application files.

=head2 lockfile

String. Default to /basedir/lockfile

=head2 _lockfile_fh

Filehandle. Used for holding the lockfile filehandle while the app is running.

=head2 pidfile

String. Default to /basedir/pid

=head1 METHODS

=head2 run

C<< run() >> will determine which command was issued to the script,
defaults to I<< start >> if no command is given. By default the valid
commands are: status, start, stop, restart, reload

New commands can be added by the consuming class, it which case the
attribute C<< _valid_commands >> needs to be updated for C<< run() >>
to allow the command to be executed. C<< _valid_commands >> is a
RegexpRef and the default value is:
C<< qr/\A(status|start|stop|reload|restart)\z/ >>

You can set your own C<< _valid_commands >> in the consuming class, to allow 
for custom commands like this:

  has '+_valid_commands' => (
    default => sub {
      return qr/\A(status|start|stop|restart|custom_command)\z/xms
    },
  );

=head2 status

Checks if the app is running, print status to STDOUT.

=head2 start

C<< start() >> call C<< main() >> after checking that it is not running
and after forking (unless foreground mode is enabled).

=head2 stop

C<< stop() >> issues a C<< TERM >> signal to the PID listed in the pidfile. It
is up to the author to trap this signal and end the application in an
orderly fashion.

=head2 restart

C<< restart() >> simply call C<< stop() >>, wait for the app to stop
and call C<< start() >>.

=head2 reload

C<< reload() >> issues a C<< HUP >> signal to the PID listed in the pidfile.
It is up to the author to trap this signal and do the appropriate
thing, usualy to reload configuration files.

=for Pod::Coverage LOCK_EX

=for Pod::Coverage LOCK_NB

=head1 AUTHOR

Tore Andersson <tore.andersson@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Tore Andersson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
